xmppserver/src/test/java/org/jivesoftware/openfire/session/ExpectedOutcome.java

```diff
@@ -0,0 +1,298 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.openfire.session;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.jivesoftware.openfire.session.ExpectedOutcome.ConnectionState.*;
+
+public class ExpectedOutcome
+{
+    /**
+     * Enumeration of possible connection states.
+     */
+    public enum ConnectionState
+    {
+        /**
+         * Connection cannot be established. In the current implementation, this includes scenarios in which connections could not be authenticated.
+         */
+        NO_CONNECTION("(no conn)"),
+
+        /**
+         * Connection without encryption, Initiating Entity is authenticated by the Receiving Entity using the Dialback protocol.
+         */
+        NON_ENCRYPTED_WITH_DIALBACK_AUTH("PLAIN DIALB"),
+
+        /**
+         * Connection that is encrypted, Initiating Entity is authenticated by the Receiving Entity using the Dialback protocol.
+         */
+        ENCRYPTED_WITH_DIALBACK_AUTH("TLS DIALB"),
+
+        /**
+         * Connection that is encrypted, Initiating Entity is authenticated by the Receiving Entity using the SASL EXTERNAL mechanism.
+         */
+        ENCRYPTED_WITH_SASLEXTERNAL_AUTH("SASL_EXT");
+
+        final String shortCode;
+        ConnectionState(String shortCode) {
+            this.shortCode = shortCode;
+        }
+
+        public String getShortCode() {
+            return shortCode;
+        }
+    }
+
+    private final Set<String> rationales = new HashSet<>();
+    private ConnectionState connectionState;
+
+    public void set(final ConnectionState state, final String rationale)
+    {
+        if (connectionState != null && connectionState != state) {
+            throw new IllegalStateException("Cannot set state " + state + ". State already is " + connectionState);
+        }
+        connectionState = state;
+        rationales.add(rationale);
+    }
+
+    public ConnectionState getConnectionState()
+    {
+        return connectionState;
+    }
+
+    public Set<String> getRationales()
+    {
+        return rationales;
+    }
+
+    public boolean isInconclusive()
+    {
+        return connectionState == null;
+    }
+
+    /**
+     * Given the configuration of the initiating and receiving server, returns the expected outcome of an outbound
+     * server-to-server connection attempt.
+     *
+     * @param initiatingServer Configuration of the local server
+     * @param receivingServer Configuration of the remote server
+     * @return the expected outcome
+     */
+    public static ExpectedOutcome generateExpectedOutcome(final ServerSettings initiatingServer, final ServerSettings receivingServer) {
+        final ExpectedOutcome expectedOutcome = new ExpectedOutcome();
+
+        switch (initiatingServer.encryptionPolicy) {
+            case disabled: // <-- Initiating server's encryption policy.
```

Comment:
Do we still have the sources for all of these decisions?
I'd love to annotate where we got them, e.g. RFC sections.
Future Us will be grateful.

Reply:
There's also a lot of TODOs and conversations in the comments. I'd be keen to get those revalidated and squashed as much as we can.

Reply:
Many of the TODOs relate to one of two things:
- Possible usage of `anon` cypher suites, that would explicitly allow TLS for encryption but not authentication. Although theoretically possible, we've never worked with this or explicitly supported it. I don't know of this to be used 'in the wild'. I've left these TODOs in the code.
- Should we allow TLS encryption (not authentication) when an invalid certificate was used? @AlexGidman introduced a `strictCertificateValidation` setting, that can be used to explicitly configure Openfire either way. In the `ExpectedOutcome` I'm now supporting that setting, removing many of the TODOs. Note that the actual test execution always use `true` value for this setting. It is hard to configure the setting to a different value for each endpoint of the conversation.

I've also added some references to documentation where the calculated decisions are made of.